### DISCUSSION TITLE
run two tasks together

### DISCUSSION THREAD

**Discussion by noahzn** [bitcraze/discussions#748] [Q&A] [answered]
Hi, I want to run two tasks on the drone together. One is the WIFI streamer demo, and the other is a custom task sending messages to cfclient.

So I defined two tasks as follows.
```
xTask = xTaskCreate(camera_task, "camera_task", ...);
xTask = xTaskCreate(my_task, "my_task", ...);

while (1)
{
pi_yield();
}

```
After flashing the code, I got nothing from the wifi streamer (python opencv-viewer.py), but some intermediate messages were printed in the cfclient. If I delete the second task defined above, the wifi streamer worked. My question is how can I run two tasks together? Thank you.

**Comment by krichardsson** [comment, marked as answer]
Hi,
I assume you are talking about code in the GAP8?

Generally there is no problem starting and running 2 tasks at the same time. You would do something similar to:
```C
void camera_task(void *parameters) {
// Do stuff
}
void my_task(void *parameters) {
// Do other stuff
}

xTask = xTaskCreate(camera_task, "camera_task", ...);
xTask = xTaskCreate(my_task, "my_task", ...);

while (1)
{
pi_yield();
}
```
The two task start and conceptually they will run in parallel, in reality the CPU can only execute one thing at a time and the OS will switch between the tasks in the system. Switching happens when a task goes to sleep, waits for a queue or similar operations. It also happens every millisecond when the OS re-schedules the tasks. A well-behaved task should not run all the time, it should hand over the possibility for other tasks to execute

```C
// Do not do this
void bad_task(void *parameters) {
  int count = 0;
  while(true) {
    // Spin-loop, using all available CPU and starving other tasks
    count++;
  }
}

// Do this
void good_task(void *parameters) {
  int count = 0;
  while(true) {
    count++;
    // Let other tasks execute
    vTaskDelay(100);
  }
}
```

Is it maybe possible that you use a lot of CPU in the `my_task`?

> My question is how can I run two tasks together?

I'm not exactly sure what you mean with "together", but some care must be taken when sharing data between tasks as you never know exactly when a context switch will occur. You can read about the tools available in [Free-RTOS inter-task communication in the documentation.](https://freertos.org/Inter-Task-Communication.html)

**Comment by noahzn** [comment]
Hi @krichardsson, thank you for your reply. `my_task` defines the running of a CNN and sending messages (commands) from GAP8 to the APP layer. But currently, I haven't integrated the CNN code. I only want to test the running of two tasks together. So my final expectation is:
1. the camera captures the image and  the image is sent to the laptop using WIFI (+ `fov.py`);
2. the same image is also sent to the CNN for outputting a number;
3.  the function sends this number to the APP layer to control the drone.
4. the camera captures the next frame and repeats 1-3.
Is this possible? Thank you in advance.

**Reply by krichardsson** [reply]
That sounds pretty straight forward, I think you can look at the [FaceDetection example](https://github.com/bitcraze/aideck-gap8-examples/tree/master/examples/image_processing/FaceDetection) for inspiration, there are a lot of similarities.

We recently [merged a PR](https://github.com/bitcraze/crazyflie-firmware/pull/1279) that should make it easier to communicate with an app in the STM from the ESP or GAP8.

**Reply by noahzn** [reply]
Thank you! Also the merged PR is useful.
